src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -794,25 +790,35 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
                     builder.ElementTypeWithAnnotations = collectionType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.Single();
 
                     MethodSymbol getEnumeratorMethod = GetGetEnumeratorMethod(collectionType, diagnostics, isAsync, errorLocationSyntax);
+                    MethodSymbol moveNextMethod = null;
 
                     if ((object)getEnumeratorMethod != null)
                     {
                         MethodSymbol specificGetEnumeratorMethod = getEnumeratorMethod.AsMember(collectionType);
-                        builder.GetEnumeratorMethod = specificGetEnumeratorMethod;
                         TypeSymbol enumeratorType = specificGetEnumeratorMethod.ReturnType;
 
                         if (isAsync)
                         {
                             Debug.Assert(enumeratorType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)));
+                            Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1);
+                            Debug.Assert(specificGetEnumeratorMethod.Parameters[0].IsOptional);
+
+                            // IAsyncEnumerable<T>.GetAsyncEnumerator has a default param, so let's fill it in
+                            builder.GetEnumeratorInfo = BindDefaultArguments(specificGetEnumeratorMethod, extensionReceiverOpt: null, allowExtensionMethods: false, expanded: false, collectionExpr.Syntax, diagnostics);
 
                             MethodSymbol moveNextAsync = (MethodSymbol)GetWellKnownTypeMember(Compilation, WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__MoveNextAsync,
                                 diagnostics, errorLocationSyntax.Location, isOptional: false);
 
                             if ((object)moveNextAsync != null)
                             {
-                                builder.MoveNextMethod = moveNextAsync.AsMember((NamedTypeSymbol)enumeratorType);
+                                moveNextMethod = moveNextAsync.AsMember((NamedTypeSymbol)enumeratorType);
                             }
                         }
+                        else
+                        {
+                            // We know that IEnumerable<T>.GetEnumerator has no parameters
```

Comment:
>// We know that IEnumerable<T>.GetEnumerator has no parameters [](start = 28, length = 62)

Similar questions here. I think that rather than making assumptions like that we should simply have a helper that looks at the signature and does the right thing. based on the method in hands.